cpp/src/arrow/csv/writer.cc

```diff
@@ -328,6 +359,31 @@ class QuotedColumnPopulator : public ColumnPopulator {
   }
 
  private:
+  // Returns true if there's no quote in the string array
+  // similar to std::find, but with much better performance
+  static bool NoQuoteInArray(const StringArray& array) {
+    const uint8_t* const data = array.raw_data() + array.value_offset(0);
+    const int64_t buffer_size = array.total_values_length();
+    int64_t offset = 0;
+#if defined(ARROW_HAVE_SSE4_2) || defined(ARROW_HAVE_NEON)
+    using simd_batch = xsimd::make_sized_batch_t<uint8_t, 16>;
+    while ((offset + 16) <= buffer_size) {
+      const auto v = simd_batch::load_unaligned(data + offset);
```

Comment:
should we have a predessor loop that loops until alignment so we can then do aligned values here?

Reply:
Have we defined standard for using XSIMD?  We can achieve similar results with slightly modified code: https://godbolt.org/z/qeM3YEYzK

Reply:
https://godbolt.org/z/o3dWdrqqz makes use of count quote.

Reply:
No obvious difference for aligned access per my test, probably because this code is not the hotspot.
That said, unaligned loading does have observable penalty for trivial code, e.g., https://quick-bench.com/q/MLvhY9_3IRZK0UR66DL1KFOF0xc

We don't have standard for using xsimd.

Yes we can leverage compiler auto-vectorization to avoid explicit simd code. Like your example, compiler is able to vectorize `std::count` like code pretty well. One reason I'm using xsimd here is there's already a similar optimization in csv writer to find string cell with illegal chars: https://github.com/apache/arrow/blob/master/cpp/src/arrow/csv/writer.cc#L213-L249

Reply:
Rewritten without explicit simd code. Slightly worse than simd code (~10% for SringNoQuote/0 test), but easier to maintain.

Reply:
Maybe it would be better to have a uniform SIMD approach that can be consistently applied throughout the codebase?

Reply:
Probably would also want to support AVX, and later AVX512, SVE.

Reply:
We adopted xsimd library to unify simd code across different architectures.

Reply:
Thanks. What is needed as a standard for using xsimd?

Reply:
Just a rule of thumb on when to use XSIMD which seems to require if/defs vs code that can be auto-vectorized.  I guess its a judgement call.